### PR TITLE
Reject submitted jobs that exceed user's quota

### PR DESCRIPTION
**Pull request by DaoWen**
## Changes proposed in this PR

If a single job requires more resources than the user's entire quota, then reject the job.

## Why are we making these changes?

It seems better to reject an unschedulable job at submission time, rather than letting the user submit a job that can never be scheduled (assuming the user's quota doesn't change).

This is particularly useful in scenarios where an organization has multiple Cook clusters, but the user only has quota on some of those clusters. Currently, a user could submit a bunch of jobs to a cluster where they have no quota, and the jobs just sit there forever. After this change, all jobs would be rejected on any clusters where the user has no quota.

### CODE REVIEW

**Comment by DaoWen** [comment]
I ran into several bugs in existing code when testing this change. Added some band-aids to get the tests to pass, but there are still several outstanding issues in #691 that should be fixed.

Also, the quota unit tests were not being run because the namespace was misnamed... 😖

File: integration/tests/cook/test_multi_user.py [hunk 1]
```diff
@@ -84,3 +84,41 @@ def test_multi_user_usage(self):
...
+                self.assertEqual(resp.status_code, 422, msg=resp.text)
+            # User with tiny quota can't submit bigger jobs, but can submit tiny jobs
+            with admin:
+                resp = util.set_limit(self.cook_url, 'quota', user.name, cpus=4)
+                self.assertEqual(resp.status_code, 201, resp.text)
+            with user:
+                _, resp = util.submit_job(self.cook_url, cpus=5)
+                self.assertEqual(resp.status_code, 422, msg=resp.text)
+                _, resp = util.submit_job(self.cook_url, cpus=4)
+                self.assertEqual(resp.status_code, 201, msg=resp.text)
```
Comment (dposada) [inline integration/tests/cook/test_multi_user.py]: Maybe we should make this case even tinier so that we don't actually use 4 cpus on the cluster when we run the test

File: integration/tests/cook/test_multi_user.py [hunk 2]
```diff
@@ -84,3 +84,41 @@ def test_multi_user_usage(self):
...
+            # Reset user's quota back to default, then user can submit jobs again
+            with admin:
+                resp = util.reset_limit(self.cook_url, 'quota', user.name)
+                self.assertEqual(resp.status_code, 204, resp.text)
+            with user:
+                _, resp = util.submit_job(self.cook_url)
+                self.assertEqual(resp.status_code, 201, msg=resp.text)
+        finally:
+            with admin:
+                util.reset_limit(self.cook_url, 'quota', user.name)
```
Comment (dposada) [inline integration/tests/cook/test_multi_user.py]: I wonder if we should replicate some or all of the scenarios but with mem instead of cpus

File: integration/tests/cook/util.py [hunk 3]
```diff
@@ -869,14 +869,15 @@ def query_queue(cook_url):
     return session.get(f'{cook_url}/queue')
 
 
-def set_limit(cook_url, limit_type, user, mem=None, cpus=None, gpus=None, jobs=None, reason='testing'):
+def set_limit(cook_url, limit_type, user, mem=None, cpus=None, gpus=None, count=None, reason='testing'):
```
Comment (dposada) [inline integration/tests/cook/util.py]: Seems like a docstring would be helpful here

File: scheduler/src/cook/mesos/api.clj [hunk 4]
```diff
@@ -108,9 +108,14 @@
 (def ZeroInt
   (s/both s/Int (s/pred zero? 'zero?)))
 
-(def PosNum
+(s/defschema PosNum
```
Comment (dposada) [inline scheduler/src/cook/mesos/api.clj]: What effect does `s/defschema` have?

File: scheduler/src/cook/mesos/api.clj [hunk 5]
```diff
@@ -1291,6 +1296,33 @@
       true
       [false {::error "You are not authorized to create jobs"}])))
 
+(defn no-job-exceeds-quota?
+  "Check if any of the submitted jobs exceed the user's total quota,
+   in which case the job would never be schedulable (unless the quota is increased).
+
+   If none of the jobs individually seem to exceed the user's quota, returns true;
+   otherwise, the following error structure is returned:
```
Comment (dposada) [inline scheduler/src/cook/mesos/api.clj]: If the jobs are not grouped, should we only block the individual jobs that exceed quota?

File: scheduler/src/cook/mesos/api.clj (hunk 4 above)
Comment (DaoWen) [reply scheduler/src/cook/mesos/api.clj]: I think it attaches some metadata that's checked when generating the Swagger HTML page. We should probably change all of our schema definitions to use this macro.

File: scheduler/src/cook/mesos/api.clj (hunk 5 above)
Comment (DaoWen) [reply scheduler/src/cook/mesos/api.clj]: I thought about that, but I concluded that the return status would be too messy in that case. Either we have to do something like a multi-status return code (HTTP 207 Multi-Status), or return a status that's misleading about part of the request (either we say it failed when some jobs *were* successfully submitted, or we say it was successful when in fact some jobs were rejected). Simply rejecting the whole request outright seemed significantly cleaner than either of those options.

File: integration/tests/cook/util.py (hunk 3 above)
Comment (DaoWen) [reply integration/tests/cook/util.py]: Added

File: integration/tests/cook/test_multi_user.py (hunk 2 above)
Comment (DaoWen) [reply integration/tests/cook/test_multi_user.py]: Done

File: scheduler/src/cook/mesos/api.clj (hunk 5 above)
Comment (dposada) [reply scheduler/src/cook/mesos/api.clj]: Makes sense

File: scheduler/src/cook/mesos/api.clj [hunk 6]
```diff
@@ -1758,7 +1792,7 @@
 
 (defn set-limit-params
   [limit-type]
-  {:body-params (merge UserLimitChangeParams {limit-type {s/Keyword s/Num}})})
+  {:body-params (merge UserLimitChangeParams {limit-type {s/Keyword NonNegNum}})})
```
Comment (dposada) [inline scheduler/src/cook/mesos/api.clj]: Should we add an integration test for this?

File: scheduler/test/cook/test/mesos/api.clj [hunk 7]
```diff
@@ -554,6 +554,15 @@
...
 
+    (testing "quota is checked on job submission"
+      (let [job (assoc (basic-job) "cpus" 32.0)
+            job-resp (h (merge quota-req-attrs
+                               {:uri "/rawscheduler"
+                                :authorization/user "foo"
+                                :request-method :post
+                                :body-params {"jobs" [job]}}))]
+        (is (= 422 (:status job-resp)))))
+
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/api.clj]: Doesn't the integration test cover this? Are we getting any additional value out of the unit test?

File: scheduler/test/cook/test/mesos/quota.clj [hunk 8]
```diff
@@ -29,6 +29,8 @@
     (quota/set-quota! conn "u1" "too many CPUs" :cpus 5.0)
     (quota/set-quota! conn "u1" "higher count" :count 6)
     (quota/set-quota! conn "u2" "custom limits" :cpus 5.0  :mem 10.0)
+    (quota/set-quota! conn "u3" "needs no GPUs" :gpus 0.0)
+    (quota/set-quota! conn "u4" "no jobs allowed" :count 0)
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/quota.clj]: Why were these added? Do we need additional assertions for this?

File: scheduler/test/cook/test/mesos/api.clj (hunk 7 above)
Comment (DaoWen) [reply scheduler/test/cook/test/mesos/api.clj]: I added this because I was debugging locally and this was much faster than doing integration tests, but now that it's stable it's just redundant. I'll remove it.

Comment (DaoWen) [reply scheduler/test/cook/test/mesos/api.clj]: I deleted the redundant CPUs test, but it occurred to me that we can test GPUs here (which are not being tested in the integration tests), so now this is a GPU-quota test.
